apps/remote_control/lib/lexical/remote_control/code_intelligence/ast.ex

```diff
@@ -0,0 +1,197 @@
+defmodule Lexical.RemoteControl.CodeIntelligence.Ast do
+  alias Lexical.Document
+  alias Lexical.Document.Position
+
+  def original_module(aliases_mapping, module) do
+    module_path = module |> to_string() |> String.split(".")
+
+    case module_path do
+      ["__MODULE__"] ->
+        aliases_mapping[module]
+
+      ["Elixir", as_alias | tail] ->
+        as = Module.concat([as_alias])
+        alias = Map.get(aliases_mapping, as, as)
+        Module.concat([alias | tail])
+
+      ["Elixir" | _] ->
+        Map.get(aliases_mapping, module, module)
+
+      _ ->
+        module
+    end
+  end
+
+  def aliases_at_position(%Document{} = document, %Position{} = position) do
+    document
+    |> Document.fragment(position)
+    |> collect_aliases()
+  end
+
+  defp collect_aliases(fragment) do
+    case Code.Fragment.container_cursor_to_quoted(fragment) do
+      {:ok, quoted} ->
+        quoted
+        |> alias_mapping()
+        |> Map.merge(current_module_mapping(quoted))
+
+      _ ->
+        %{}
+    end
+  end
+
+  defp alias_mapping(quoted) do
```

Comment:
does this function need to be written in terms of a reduce over the current document? 
I'm imagining you collect aliases in a context as you encounter them, and then store ranges where the context is valid. I think that'd improve the readability of a lot of this, and would prevent you from needing all these helpers that need to go back earlier in the document and rebuild context. 

If nothing else, you could still use this method to build up aliases until a certain point, you just stop at that position.